File: driver/modbus/scan_task_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "gtest/gtest.h"
#include "nlohmann/json.hpp"

#include "x/cpp/breaker/breaker.h"
#include "x/cpp/defer/defer.h"
#include "x/cpp/xtest/xtest.h"

#include "driver/modbus/mock/slave.h"
#include "driver/modbus/scan_task.h"
#include "driver/task/common/scan_task.h"

/// @brief it should successfully test connection to Modbus device.
TEST(ScanTask, testConnection) {
    auto slave = modbus::mock::Slave(modbus::mock::SlaveConfig{});
    ASSERT_NIL(slave.start());
    x::defer stop_slave([&slave] { slave.stop(); });
    auto ctx = std::make_shared<task::MockContext>(nullptr);

    synnax::Task t;
    t.key = 12345;
    t.type = "modbus_scan";

    auto dev_manager = std::make_shared<modbus::device::Manager>();

    const auto cfg = modbus::ScanTaskConfig{};
    auto scan_task = std::make_unique<common::ScanTask>(
        std::make_unique<modbus::Scanner>(ctx, t, dev_manager),
        ctx,
        t,
        breaker::default_config(t.name),
        cfg.scan_rate
    );

    auto conn_cfg = modbus::device::ConnectionConfig{"127.0.0.1", 1502};
    auto cmd_args = json{{"connection", conn_cfg.to_json()}};
    auto cmd = task::Command(t.key, modbus::TEST_CONNECTION_CMD_TYPE, cmd_args);
    cmd.key = "electric_boogaloo";

    scan_task->exec(cmd);
    ASSERT_EQ(ctx->statuses.size(), 1);
    auto first = ctx->statuses[0];
    EXPECT_EQ(first.variant, status::variant::SUCCESS);
    EXPECT_EQ(first.key, t.status_key());
    EXPECT_EQ(first.details.cmd, cmd.key);
    EXPECT_EQ(first.details.task, t.key);
    EXPECT_EQ(first.message, "Connection successful");
}

TEST(ScanTask, testConfigReturnsCorrectValues) {
    auto ctx = std::make_shared<task::MockContext>(nullptr);
    synnax::Task t;
    t.key = 12345;
    t.type = "modbus_scan";
    auto dev_manager = std::make_shared<modbus::device::Manager>();

    const modbus::Scanner scanner(ctx, t, dev_manager);
    auto cfg = scanner.config();
    EXPECT_EQ(cfg.make, "modbus");
    EXPECT_EQ(cfg.log_prefix, "[modbus.scan_task]");
}

TEST(ScanTask, testExecReturnsFalseForUnknownCommand) {
    auto ctx = std::make_shared<task::MockContext>(nullptr);
    synnax::Task t;
    t.key = 12345;
    t.type = "modbus_scan";
    auto dev_manager = std::make_shared<modbus::device::Manager>();

    modbus::Scanner scanner(ctx, t, dev_manager);
    task::Command cmd(t.key, "unknown_command", json{});
    bool handled = scanner.exec(cmd, t, ctx);
    EXPECT_FALSE(handled);
}

TEST(ScanTask, testScanChecksDeviceHealth) {
    auto slave = modbus::mock::Slave(modbus::mock::SlaveConfig{});
    ASSERT_NIL(slave.start());
    x::defer stop_slave([&slave] { slave.stop(); });

    auto ctx = std::make_shared<task::MockContext>(nullptr);
    synnax::Task t;
    t.key = 12345;
    t.type = "modbus_scan";
    auto dev_manager = std::make_shared<modbus::device::Manager>();

    modbus::Scanner scanner(ctx, t, dev_manager);

    // Create device with valid Modbus connection properties
    synnax::Device dev;
    dev.key = "health-test-device";
    dev.name = "Health Test Device";
    dev.make = "modbus";
    dev.rack = synnax::rack_key_from_task_key(t.key);
    dev.properties = json{
        {"connection",
         {{"host", "127.0.0.1"},
          {"port", 1502},
          {"swap_bytes", false},
          {"swap_words", false}}}
    }.dump();

    std::unordered_map<std::string, synnax::Device> devices_map;
    devices_map[dev.key] = dev;
    common::ScannerContext scan_ctx;
    scan_ctx.devices = &devices_map;

    auto devices = ASSERT_NIL_P(scanner.scan(scan_ctx));
    ASSERT_EQ(devices.size(), 1);
    EXPECT_EQ(devices[0].status.variant, status::variant::SUCCESS);
    EXPECT_EQ(devices[0].status.message, "Device connected");
}

TEST(ScanTask, testScanReportsDisconnectedDevice) {
    auto ctx = std::make_shared<task::MockContext>(nullptr);
    synnax::Task t;
    t.key = 12345;
    t.type = "modbus_scan";
    auto dev_manager = std::make_shared<modbus::device::Manager>();

    modbus::Scanner scanner(ctx, t, dev_manager);

    // Create device with invalid connection (no server running on this port)
    synnax::Device dev;
    dev.key = "disconnected-device";
    dev.name = "Disconnected Device";
    dev.make = "modbus";
    dev.rack = synnax::rack_key_from_task_key(t.key);
    dev.properties = json{
        {"connection",
         {{"host", "127.0.0.1"},
          {"port", 9999},
          {"swap_bytes", false},
          {"swap_words", false}}}
    }.dump();

    std::unordered_map<std::string, synnax::Device> devices_map;
    devices_map[dev.key] = dev;
    common::ScannerContext scan_ctx;
    scan_ctx.devices = &devices_map;

    auto devices = ASSERT_NIL_P(scanner.scan(scan_ctx));
    ASSERT_EQ(devices.size(), 1);
    EXPECT_EQ(devices[0].status.variant, status::variant::WARNING);
    EXPECT_EQ(devices[0].status.message, "Failed to reach device");
}

File: core/pkg/service/driver/scan_task_test.go
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package driver_test

import (
	"context"
	"sync"
	"time"

	"github.com/google/uuid"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/synnaxlabs/synnax/pkg/service/device"
	"github.com/synnaxlabs/synnax/pkg/service/driver"
	"github.com/synnaxlabs/synnax/pkg/service/rack"
	"github.com/synnaxlabs/synnax/pkg/service/status"
	"github.com/synnaxlabs/synnax/pkg/service/task"
	"github.com/synnaxlabs/x/encoding/msgpack"
	"github.com/synnaxlabs/x/errors"
	. "github.com/synnaxlabs/x/testutil"
)

// mapScanner is a driver.Scanner whose check results a spec sets per device.
type mapScanner struct {
	mu sync.Mutex
	// unreachable holds the check error of each device that is not reachable.
	unreachable map[device.Key]error
	checks      map[device.Key]int
}

func newMapScanner() *mapScanner {
	return &mapScanner{
		unreachable: make(map[device.Key]error),
		checks:      make(map[device.Key]int),
	}
}

func (s *mapScanner) setUnreachable(key device.Key, err error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err == nil {
		delete(s.unreachable, key)
		return
	}
	s.unreachable[key] = err
}

func (s *mapScanner) checkCount(key device.Key) int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.checks[key]
}

func (s *mapScanner) Check(_ context.Context, dev device.Device) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.checks[dev.Key]++
	return s.unreachable[dev.Key]
}

func (s *mapScanner) Exec(
	_ context.Context,
	cmd task.Command,
) (msgpack.EncodedJSON, error) {
	switch cmd.Type {
	case "test_connection":
		return msgpack.EncodedJSON{"latency_ms": 4.0}, nil
	case "browse":
		return nil, errors.New("broker refused the subscription")
	}
	return nil, driver.ErrUnsupportedCommand
}

var _ = Describe("ScanTask", func() {
	var (
		t       task.Task
		scanner *mapScanner
		rackKey rack.Key
		make_   string
	)

	BeforeEach(func(ctx SpecContext) {
		r := rack.Rack{Name: "scan-task-rack"}
		Expect(rackService.NewWriter(nil).Create(ctx, &r)).To(Succeed())
		rackKey = r.Key
		make_ = "make-" + uuid.NewString()
		t = task.Task{
			Key:  uuid.New(),
			Name: "scan-task-test",
			Type: "mock",
			Rack: rackKey,
		}
		scanner = newMapScanner()
	})

	open := func(interval time.Duration) *driver.ScanTask {
		GinkgoHelper()
		st := MustSucceed(driver.NewScanTask(driver.ScanTaskConfig{
			Scanner:            scanner,
			Status:             statusSvc,
			Device:             deviceSvc,
			Make:               make_,
			ReachableMessage:   "Broker connected",
			UnreachableMessage: "Failed to reach broker",
			Task:               t,
			Interval:           interval,
		}))
		DeferCleanup(func() { Expect(st.Stop(false)).To(Succeed()) })
		return st
	}

	createDevice := func(ctx context.Context, deviceMake string) device.Device {
		GinkgoHelper()
		dev := device.Device{
			Key:      uuid.NewString(),
			Rack:     rackKey,
			Location: "localhost:1883",
			Name:     "Broker",
			Make:     deviceMake,
			Model:    "broker",
		}
		Expect(deviceSvc.NewWriter(nil).Create(ctx, &dev)).To(Succeed())
		return dev
	}

	deviceStatus := func(ctx context.Context, dev device.Device) device.Status {
		var stat device.Status
		_ = statusSvc.NewRetrieve[device.StatusDetails]().
			Where(status.MatchKeys[device.StatusDetails](dev.OntologyID().String())).
			Entry(&stat).Exec(ctx, nil)
		return stat
	}

	taskStatus := func(ctx context.Context) task.Status {
		var stat task.Status
		_ = statusSvc.NewRetrieve[task.StatusDetails]().
			Where(status.MatchKeys[task.StatusDetails](t.OntologyID().String())).
			Entry(&stat).Exec(ctx, nil)
		return stat
	}

	Describe("NewScanTask", func() {
		It("Should reject a configuration with no make", func() {
			Expect(driver.NewScanTask(driver.ScanTaskConfig{
				Scanner:            scanner,
				Status:             statusSvc,
				Device:             deviceSvc,
				ReachableMessage:   "Broker connected",
				UnreachableMessage: "Failed to reach broker",
				Task:               t,
			})).Error().To(MatchError(ContainSubstring("make")))
		})
	})

	Describe("Device checks", func() {
		It("Should report each device of its make as reachable or not",
			func(ctx SpecContext) {
				up := createDevice(ctx, make_)
				down := createDevice(ctx, make_)
				other := createDevice(ctx, "other-make")
				scanner.setUnreachable(down.Key, errors.New("dial timeout"))
				st := open(time.Hour)
				Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
				Eventually(func(g Gomega) {
					stat := deviceStatus(ctx, up)
					g.Expect(stat.Variant).To(Equal(status.VariantSuccess))
					g.Expect(stat.Message).To(Equal("Broker connected"))
					g.Expect(stat.Details.Device).To(Equal(up.Key))
					g.Expect(stat.Details.Rack).To(Equal(rackKey))
				}).Should(Succeed())
				stat := deviceStatus(ctx, down)
				Expect(stat.Variant).To(Equal(status.VariantWarning))
				Expect(stat.Message).To(Equal("Failed to reach broker"))
				Expect(stat.Description).To(Equal("dial timeout"))
				Expect(scanner.checkCount(other.Key)).To(BeZero())
			},
		)

		It("Should check a device when it is created", func(ctx SpecContext) {
			st := open(time.Hour)
			Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
			dev := createDevice(ctx, make_)
			Eventually(func(g Gomega) {
				g.Expect(deviceStatus(ctx, dev).Message).To(Equal("Broker connected"))
			}).Should(Succeed())
		})

		It("Should report a change of health on the next interval",
			func(ctx SpecContext) {
				dev := createDevice(ctx, make_)
				st := open(5 * time.Millisecond)
				Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
				Eventually(func(g Gomega) {
					g.Expect(deviceStatus(ctx, dev).Variant).
						To(Equal(status.VariantSuccess))
				}).Should(Succeed())
				scanner.setUnreachable(dev.Key, errors.New("connection reset"))
				Eventually(func(g Gomega) {
					stat := deviceStatus(ctx, dev)
					g.Expect(stat.Variant).To(Equal(status.VariantWarning))
					g.Expect(stat.Description).To(Equal("connection reset"))
				}).Should(Succeed())
			},
		)

		It("Should write a status only when the health changes", func(ctx SpecContext) {
			dev := createDevice(ctx, make_)
			st := open(5 * time.Millisecond)
			Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
			Eventually(func(g Gomega) {
				g.Expect(deviceStatus(ctx, dev).Message).To(Equal("Broker connected"))
			}).Should(Succeed())
			first := deviceStatus(ctx, dev).Time
			checks := scanner.checkCount(dev.Key)
			Eventually(func() int { return scanner.checkCount(dev.Key) }).
				Should(BeNumerically(">", checks+2))
			Expect(deviceStatus(ctx, dev).Time).To(Equal(first))
		})

		It("Should stop the checks on a stop command", func(ctx SpecContext) {
			dev := createDevice(ctx, make_)
			st := open(5 * time.Millisecond)
			Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
			Eventually(func() int { return scanner.checkCount(dev.Key) }).
				Should(BeNumerically(">", 0))
			Expect(st.Exec(ctx, task.Command{Type: "stop", Key: "cmd-1"})).
				To(Succeed())
			checks := scanner.checkCount(dev.Key)
			Consistently(func() int { return scanner.checkCount(dev.Key) }, "50ms").
				Should(Equal(checks))
			stat := taskStatus(ctx)
			Expect(stat.Details.Running).To(BeFalse())
			Expect(stat.Details.Cmd).To(Equal("cmd-1"))
		})
	})

	Describe("Exec", func() {
		It("Should answer a scanner command with its data", func(ctx SpecContext) {
			st := open(time.Hour)
			Expect(st.Start(ctx, driver.NoCommand)).To(Succeed())
			Expect(st.Exec(ctx, task.Command{Type: "test_connection", Key: "cmd-2"})).
				To(Succeed())
			stat := taskStatus(ctx)
			Expect(stat.Variant).To(Equal(status.VariantSuccess))
			Expect(stat.Details.Cmd).To(Equal("cmd-2"))
			Expect(stat.Details.Running).To(BeTrue())
			Expect(stat.Details.Data).To(HaveKeyWithValue("latency_ms", 4.0))
		})

		It("Should answer a failed scanner command with its error",
			func(ctx SpecContext) {
				st := open(time.Hour)
				Expect(st.Exec(ctx, task.Command{Type: "browse", Key: "cmd-3"})).
					To(MatchError(ContainSubstring("broker refused the subscription")))
				stat := taskStatus(ctx)
				Expect(stat.Variant).To(Equal(status.VariantError))
				Expect(stat.Message).To(Equal("broker refused the subscription"))
				Expect(stat.Details.Cmd).To(Equal("cmd-3"))
			},
		)

		It("Should reject a command the scanner does not support",
			func(ctx SpecContext) {
				st := open(time.Hour)
				Expect(st.Exec(ctx, task.Command{Type: "tare", Key: "cmd-4"})).
					To(MatchError(driver.ErrUnsupportedCommand))
				Expect(taskStatus(ctx).Details.Cmd).ToNot(Equal("cmd-4"))
			},
		)
	})
})
